api/runtime/hooks/v1alpha1/upgrade_plan_types.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1beta1 "sigs.k8s.io/cluster-api/api/core/v1beta1"
+	runtimecatalog "sigs.k8s.io/cluster-api/exp/runtime/catalog"
+)
+
+// GenerateUpgradePlanRequest is the request of the GenerateUpgradePlan hook.
+// +kubebuilder:object:root=true
+type GenerateUpgradePlanRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonRequest contains fields common to all request types.
+	CommonRequest `json:",inline"`
+
+	// cluster is the cluster ofject the lifecycle hook correspods to.
+	// +required
+	Cluster clusterv1beta1.Cluster `json:"cluster,omitempty,omitzero"`
+
+	// fromControlPlaneKubernetesVersion is the current Kubernetes version of the control plane.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	FromControlPlaneKubernetesVersion string `json:"fromControlPlaneKubernetesVersion,omitempty"`
+
+	// fromWorkersKubernetesVersion is the current Kubernetes version of the workers.
+	// +optional
+	// +kubebuilder:validation:MinLength=1
+	FromWorkersKubernetesVersion string `json:"fromWorkersKubernetesVersion,omitempty"`
+
+	// toKubernetesVersion is the target Kubernetes version for the upgrade.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	ToKubernetesVersion string `json:"toKubernetesVersion,omitempty"`
+}
+
+var _ ResponseObject = &GenerateUpgradePlanResponse{}
+
+// GenerateUpgradePlanResponse is the response of the GenerateUpgradePlan hook.
+// +kubebuilder:object:root=true
+type GenerateUpgradePlanResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains Status and Message fields common to all response types.
+	CommonResponse `json:",inline"`
+
+	// controlPlaneUpgrades is the list of version upgrade steps for the control plane.
+	// Each entry represents an intermediate version that must be applied in sequence.
```

Comment:
Let's add some more details on the rules that applies to the upgrade plan
```suggestion
	// controlPlaneUpgrades is the list of version upgrade steps for the control plane.
	// Each entry represents an intermediate version that must be applied in sequence.
	// The following rules apply:
	// - there should be at least one version for every minor between 		fromControlPlaneKubernetesVersion (excluded) and ToKubernetesVersion (included).
	// - each version must be:
	//   - greater than fromControlPlaneKubernetesVersion (or with a different build 	number)
	//   - greater than the previous version in the list (or with a different build number)
	//   - less or equal to ToKubernetesVersion (or with a different build number)
	//   - the last version in the plan must be equal to ToKubernetesVersion
```